vllm/worker/vineyard_llm_cache.py

```diff
@@ -369,9 +384,11 @@ def update_seq_kv_caches(
         start_unload = torch.cuda.Event(enable_timing=True)
         end_unload = torch.cuda.Event(enable_timing=True)
         start_unload.record()
+        # using a cuda staging buffer to avoid the inefficiency of non-contiguous HBM->DRAM memcpy
         for j in range(self.layer):
-            self.buffer[:, j, :update_token_size].copy_(
+            self.cuda_buffer[:, j, :update_token_size].copy_(
                 kv_caches[j][:, slot_mapping // block_size, slot_mapping % block_size])
+        self.buffer.copy_(self.cuda_buffer)
```

Comment:
I kind of forget why do we copy from `buffer` to `cuda_buffer` here?

Reply:
this is a copy from `cuda_buffer` to `buffer`, which converts original non-contiguous HBM->DRAM copy (kv_caches (non contiguous) -> buffer) to contiguous copy (kv_caches (non contiguous) -> cuda_buffer (contiguous) -> buffer)